viewer/src/datamodels/cad/rendering/TransformOverrideBuffer.ts

```diff
@@ -0,0 +1,115 @@
+/*!
+ * Copyright 2020 Cognite AS
+ */
+
+import * as THREE from 'three';
+import { packFloat } from '@/utilities/packFloatToVec4';
+import { determinePowerOfTwoDimensions } from '@/utilities/determinePowerOfTwoDimensions';
+
+export class TransformOverrideBuffer {
+  private readonly MIN_NUMBER_OF_TREE_INDECES = 16;
+  private readonly NUMBER_OF_ELEMENTS_PER_MATRIX = 16;
+  private readonly BYTES_PER_FLOAT = 4;
+
+  private _dataTexture: THREE.DataTexture;
+  private _textureBuffer: Uint8Array;
+
+  private _unusedIndices: number[];
+
+  private _treeIndexToOverrideIndex: Map<number, number>;
+
+  private _onGenerateNewDataTextureCallback: (datatexture: THREE.DataTexture) => void;
+
+  get dataTexture(): THREE.DataTexture {
+    return this._dataTexture;
+  }
+
+  get overrideIndices() {
+    return this._treeIndexToOverrideIndex;
+  }
+
+  constructor(onGenerateNewDataTexture: (datatexture: THREE.DataTexture) => void) {
+    this._textureBuffer = new Uint8Array(
+      this.MIN_NUMBER_OF_TREE_INDECES * this.NUMBER_OF_ELEMENTS_PER_MATRIX * this.BYTES_PER_FLOAT
+    );
+
+    this._dataTexture = new THREE.DataTexture(
+      this._textureBuffer,
+      this.NUMBER_OF_ELEMENTS_PER_MATRIX,
+      this.MIN_NUMBER_OF_TREE_INDECES
+    );
+
+    this._onGenerateNewDataTextureCallback = onGenerateNewDataTexture;
+
+    this._unusedIndices = [...Array(this.MIN_NUMBER_OF_TREE_INDECES).keys()].map((_, n) => n);
+
+    this._treeIndexToOverrideIndex = new Map();
+  }
+
+  public addOverrideTransform(treeIndex: number, transform: THREE.Matrix4): number {
+    const transformBuffer = transform.toArray();
+
+    let matrixIndex: number | undefined;
+
+    if (this._treeIndexToOverrideIndex.has(treeIndex)) {
+      matrixIndex = this._treeIndexToOverrideIndex.get(treeIndex)!;
+    } else {
+      matrixIndex = this._unusedIndices.pop();
+
+      if (matrixIndex === undefined) {
+        this.reComputeDataTexture();
+        matrixIndex = this._unusedIndices.pop()!;
+      }
+
+      this._treeIndexToOverrideIndex.set(treeIndex, matrixIndex);
+    }
+
+    for (let i = 0; i < this.NUMBER_OF_ELEMENTS_PER_MATRIX; i++) {
+      const element = packFloat(transformBuffer[(i % 4) * 4 + Math.floor(i / 4)]);
+
+      const byteIndex = (matrixIndex * this.NUMBER_OF_ELEMENTS_PER_MATRIX + i) * this.BYTES_PER_FLOAT;
+
+      this._dataTexture.image.data[byteIndex] = element.x;
+      this._dataTexture.image.data[byteIndex + 1] = element.y;
+      this._dataTexture.image.data[byteIndex + 2] = element.z;
+      this._dataTexture.image.data[byteIndex + 3] = element.w;
```

Comment:
Suggestion: make `packFloatInto(value, targetBuffer, offset)` that packs the values directly into a preallocated buffer to avoid `new THREE.Vector4` and copy of values.